consensus/polybft/extra.go

```diff
@@ -142,14 +142,81 @@ func (i *Extra) UnmarshalRLPWith(v *fastrlp.Value) error {
 	return nil
 }
 
-// ValidateBasic contains extra data basic set of validations
-func (i *Extra) ValidateBasic(parentExtra *Extra) error {
+// ValidateFinalizedHeader contains extra data validations for finalized headers
+func (i *Extra) ValidateFinalizedHeader(header *types.Header, parent *types.Header, parents []*types.Header,
+	chainID uint64, consensusBackend polybftBackend, logger hclog.Logger) error {
+	// validate committed signatures
+	blockNumber := header.Number
+	if i.Committed == nil {
+		return fmt.Errorf("failed to verify signatures for block %d, because signatures are not present", blockNumber)
+	}
+
+	if i.Checkpoint == nil {
+		return fmt.Errorf("failed to verify signatures for block %d, because checkpoint data are not present", blockNumber)
+	}
+
+	checkpointHash, err := i.Checkpoint.Hash(chainID, header.Number, header.Hash)
+	if err != nil {
+		return fmt.Errorf("failed to calculate proposal hash: %w", err)
+	}
+
+	validators, err := consensusBackend.GetValidators(blockNumber-1, parents)
+	if err != nil {
+		return fmt.Errorf("failed to validate header for block %d. could not retrieve block validators:%w", blockNumber, err)
+	}
+
+	if err := i.Committed.Verify(validators, checkpointHash, logger); err != nil {
+		return fmt.Errorf("failed to verify signatures for block %d (proposal hash %s): %w",
+			blockNumber, checkpointHash, err)
+	}
+
+	parentExtra, err := GetIbftExtra(parent.ExtraData)
+	if err != nil {
+		return fmt.Errorf("failed to verify signatures for block %d: %w", blockNumber, err)
+	}
+
+	// validate parent signatures
+	if err := i.ValidateParentSignatures(blockNumber, consensusBackend, parents,
+		parent, parentExtra, chainID, logger); err != nil {
+		return err
+	}
+
 	return i.Checkpoint.ValidateBasic(parentExtra.Checkpoint)
 }
 
-// Validate contains extra data validation logic
-func (i *Extra) Validate(parentExtra *Extra, currentValidators AccountSet, nextValidators AccountSet) error {
-	return i.Checkpoint.Validate(parentExtra.Checkpoint, currentValidators, nextValidators)
+// ValidateParentSignatures validates signatures for parent block
+func (i *Extra) ValidateParentSignatures(blockNumber uint64, consensusBackend polybftBackend, parents []*types.Header,
+	parent *types.Header, parentExtra *Extra, chainID uint64, logger hclog.Logger) error {
+	// skip block 1 because genesis does not have committed signatures
+	if blockNumber <= 1 {
+		return nil
+	}
+
+	if i.Parent == nil {
+		return fmt.Errorf("failed to verify signatures for parent of block %d because signatures are not present",
+			blockNumber)
+	}
+
+	parentValidators, err := consensusBackend.GetValidators(blockNumber-2, parents)
```

Comment:
Why do you want to check validators for  blockNumber - 2?

Reply:
## Long story short...
because we want to skip epoch-ending blocks, in order to retrieve the correct validator set.

## Example
Epoch size=10 blocks
We are currently validating block#11, which means that we need to check its parent signatures (signatures of validators that have sealed block#10). Since block#10 is an epoch-ending block, it contains the `commitEpoch` transaction (i.e. validator set transition). If we were about to retrieve the validator set on block#10, we would get a validator set for the next epoch. However, we don't expect validators for the next epoch, but the ones which sealed the epoch-ending block itself (which are the validators saved for block#9 or any other block within the epoch). So that's why we are substracting by 2 here, to make sure that we have skipped an epoch-ending block and query validator set from the block within the (previous) epoch.